runtime/include/atomics/intrinsics/chpl-atomics.h

```diff
@@ -195,25 +194,28 @@ static inline chpl_bool atomic_is_lock_free_ ## type(atomic_ ## type * obj) { \
 } \
 static inline void atomic_init_ ## type(atomic_ ## type * obj, type value) { \
   assert(sizeof(type) == sizeof(uinttype)); \
-  *obj = value; \
+  memcpy(obj, &value, sizeof(*obj)); \
 } \
 static inline void atomic_destroy_ ## type(atomic_ ## type * obj) { \
 } \
 static inline void atomic_store_explicit_ ## type(atomic_ ## type * obj, type value, memory_order order) { \
-  uinttype ret_uint = *(uinttype *)obj; \
-  uinttype val_uint = *(uinttype *)&value; \
+  uinttype ret_uint = *obj; \
+  uinttype val_uint; \
   uinttype old_uint; \
+  memcpy(&val_uint, &value, sizeof(val_uint)); \
   do { \
     old_uint = ret_uint; \
-    ret_uint = __sync_val_compare_and_swap((uinttype *)obj, old_uint, val_uint); \
+    ret_uint = __sync_val_compare_and_swap(obj, old_uint, val_uint); \
   } while (ret_uint != old_uint); \
 } \
 static inline void atomic_store_ ## type(atomic_ ## type * obj, type value) { \
   atomic_store_explicit_ ## type(obj, value, memory_order_seq_cst); \
 } \
 static inline type atomic_load_explicit_ ## type(atomic_ ## type * obj, memory_order order) { \
-  full_memory_barrier(); \
-  return *obj; \
+  type ret; \
+  uinttype ret_uint = __sync_fetch_and_or(obj, (uinttype)0); \
```

Comment:
This is safe since we know it'll always be operating on a uint and never a bool, but I wonder if we should do ` __sync_val_compare_and_swap(obj, (basetype)0, (basetype)0); ` just for consistency with the integer load